### PR TITLE
Boolean, Numeric, and Structured CLI inputs

### LINKED ISSUE #4799: using `--input` with a Boolean does not work. Always thinks the value is a string
Not sure if this is a Thor issue at the core - but that would be my guess

```
inspec exec git://github.com/simp/inspec-profile-disa_stig-el7.git -t ssh://vagrant@127.0.0.1:2222 -i $SSH_KEY --reporter cli json:test1.json --input disable_slow_controls=1
Input 'disable_slow_controls' with value '1' does not validate to type 'Boolean'.
inspec exec git://github.com/simp/inspec-profile-disa_stig-el7.git -t ssh://vagrant@127.0.0.1:2222 -i $SSH_KEY --reporter cli json:test1.json --input disable_slow_controls=true
Input 'disable_slow_controls' with value 'true' does not validate to type 'Boolean'.
➜  disa_stig-el7-hardening git:(master) ✗ inspec exec git://github.com/simp/inspec-profile-disa_stig-el7.git -t ssh://vagrant@127.0.0.1:2222 -i $SSH_KEY --reporter cli json:test1.json --input disable_slow_controls='true'
Input 'disable_slow_controls' with value 'true' does not validate to type 'Boolean'.
➜  disa_stig-el7-hardening git:(master) ✗ inspec exec git://github.com/simp/inspec-profile-disa_stig-el7.git -t ssh://vagrant@127.0.0.1:2222 -i $SSH_KEY --reporter cli json:test1.json --input disable_slow_controls=1
Input 'disable_slow_controls' with value '1' does not validate to type 'Boolean'.
➜  disa_stig-el7-hardening git:(master) ✗ inspec exec git://github.com/simp/inspec-profile-disa_stig-el7.git -t ssh://vagrant@127.0.0.1:2222 -i $SSH_KEY --reporter cli json:test1.json --input disable_slow_controls true
ERROR: An '=' is required when using --input. Usage: --input input_name1=input_value1 input2=value2
➜  disa_stig-el7-hardening git:(master) ✗ inspec exec git://github.com/simp/inspec-profile-disa_stig-el7.git -t ssh://vagrant@127.0.0.1:2222 -i $SSH_KEY --reporter cli json:test1.json --input disable_slow_controls=true
Input 'disable_slow_controls' with value 'true' does not validate to type 'Boolean'.
➜  disa_stig-el7-hardening git:(master) ✗ inspec exec git://github.com/simp/inspec-profile-disa_stig-el7.git -t ssh://vagrant@127.0.0.1:2222 -i $SSH_KEY --reporter cli json:test1.json --input 'disable_slow_controls=true'
Input 'disable_slow_controls' with value 'true' does not validate to type 'Boolean'.
➜  disa_stig-el7-hardening git:(master) ✗ inspec exec git://github.com/simp/inspec-profile-disa_stig-el7.git -t ssh://vagrant@127.0.0.1:2222 -i $SSH_KEY --reporter cli json:test1.json --input disable_slow_controls=`true`
```

Aha! Link: https://chef.aha.io/features/SH-2165

### LINKED ISSUE #4963: Array inputs don't seem to be loaded correctly via the cli and `--inputs`
Arrays support seems to be working via `inspec.yml` or `input-files` but not the CLI

Thor does not seem to be respecting something like:

`--input foo=[1,2,3]` 

or something even more complex as:

`--input aoh=[{:Subject: "CN=DoD Root CA 2, OU=PKI, OU=DoD, O=U.S. Government, C=US",:Issuer: "CN=US DoD CCEB Interoperability Root CA 1, OU=PKI, OU=DoD, O=U.S. Government, C=US", :Thumbprint: "DA36FAF56B2F6FBA1604F5BE46D864C9FA013BA3", :NotAfter: "Saturday, March 09, 2019"}]`

Aha! Link: https://chef.aha.io/features/SH-2166

### LINKED ISSUE #4977: Numeric input set at CLI is treated as a string
See our tutorial on inputs:
https://www.inspec.io/docs/reference/inputs/

Declare a numeric input:

```
inputs:
- name: amplifier_max_volume
  description: How loud the amplifiers can go
  type: numeric
```

Write a failing test:

```
input('amplifier_max_volume', value: 10)

control 'Big Rock Show' do
  describe input('amplifier_max_volume') do    # This line reads the value of the input
    it { should eq 11 } # The UK'S LOUDEST BAND
  end
end
```

Make the test pass by passing a compliant value:

```
inspec exec rock_critic --input amplifier_max_volume=11
```

Expected:

From our website

```
11
     ✔  should eq 11

Test Summary: 1 successful, 0 failures, 0 skipped
```

Actual:

```
Profile: InSpec Profile (rock_show)
Version: 0.1.0
Target:  local://

  ×  Big Rock Show: 11
     ×  11 is expected to eq 11

     expected: 11
          got: "11"

     (compared using ==)
```

A workaround is to use `cmp` instead of `eq` - but we at least need to update the site for this. It should ideally be fixed because it could cause subtle and frustrating issues.

Aha! Link: https://chef.aha.io/features/SH-2164

### PR DESCRIPTION
**Pull request by clintoncwolfe**
## Description

Parses the value passed in from the `--input` CLI flag and makes some guesses as to what it could be. Maybe boolean? Maybe numeric? Maybe YAML? Maybe JSON?

Most of the time, it will fall through to being a string. There is a chance we could now wrongly throw errors on strings that begin and end with [] or {} and fail to parse as JSON or YAML, but that seems pretty edge-casey.

Tests and docs included.

## Related Issue

Fixes #4799 boolean case
Fixes #4977 numeric case
Fixes #4963 arrays, hashes and deeper things.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.

Aha! Link: https://chef.aha.io/features/SH-2196

### CODE REVIEW

File: docs/inputs.md [hunk 1]
```diff
@@ -263,6 +263,19 @@ To set multiple inputs, say:
 inspec exec my_profile --input input_name1=input_value1 name2=value2
 ```
 
+If a CLI input value resembles a number, it will be converted to an Integer or Float. Scientific notation is not currently recognized.
+```yaml
+inspec exec my_profile --input integer_input=-11
+inspec exec my_profile --input float_input=11.0
+```
+
+You may set inputs with complex values, such as arrays and hashes using either YAML or JSON syntax. Just be sure to protect the string from the shell using single quotes.
```
Comment (james-stocks) [inline docs/inputs.md:272]: Nice!

File: lib/inspec/input_registry.rb [hunk 2]
```diff
@@ -166,15 +166,44 @@ def bind_inputs_from_cli_args(profile_name, input_list)
...
+      elsif value =~ /^-?\d+\.\d+$/
+        value = value.to_f
+      elsif value =~ /^(\[|\{).*(\]|\})$/
+        # Look for complex values and try to parse them.
+        require "yaml"
+        begin
+          value = YAML.load(value)
+        rescue Psych::SyntaxError => yaml_error
+          # It could be that we just tried to run JSON through the YAML parser.
+          require "json"
```
Comment (james-stocks) [inline lib/inspec/input_registry.rb:196]: I've found that `yaml` is able to parse JSON strings in almost all cases. It's correct to do this 2nd attempt with `json` but I think very few cases will error for YAML but succeed for JSON.

File: test/fixtures/profiles/inputs/cli/controls/cli.rb [hunk 3]
```diff
@@ -8,3 +8,65 @@
     it { should cmp "value_from_cli_02"}
   end
 end
+
+control "test_control_numeric_implicit" do
+  describe input("test_input_03", value: "value_from_dsl") do
+    it { should eq 11 } # Note eq not cmp
```
Comment (james-stocks) [inline test/fixtures/profiles/inputs/cli/controls/cli.rb]: We could maybe include assertions like `its('class') { should eq Integer }` because `11` and `11.0` are different classes but return true for `eq`

File: lib/inspec/input_registry.rb [hunk 4]
```diff
@@ -166,15 +166,44 @@ def bind_inputs_from_cli_args(profile_name, input_list)
...
+          value: input_value,
           provider: :cli,
           priority: 50
         )
         find_or_register_input(input_name, profile_name, event: evt)
       end
     end
 
+    # Remove trailing commas, resolve type.
+    def clean_up_cli_input_value(input_name, given_value)
```
Comment (james-stocks) [inline lib/inspec/input_registry.rb]: Suggest `value_from_cli_input` or `parse_cli_input_value` or some other name that expresses that this method evaluates a value - the current `clean_up_` name suggests it just trims or does simple casting.

File: docs/inputs.md [hunk 5]
```diff
@@ -263,6 +263,19 @@ To set multiple inputs, say:
 inspec exec my_profile --input input_name1=input_value1 name2=value2
 ```
 
+If a CLI input value resembles a number, it will be converted to an Integer or Float. Scientific notation is not currently recognized.
+```yaml
+inspec exec my_profile --input integer_input=-11
```
Comment (james-stocks) [inline docs/inputs.md]: These examples are potentially confusing because the inputs are named by their types - this _might_ read like we are typing the inputs here. Suggest using pretend real names like `amplifier_volume` `letters_array` `fruits_json`

File: test/helper.rb [hunk 6]
```diff
@@ -148,7 +148,7 @@ def skip_until(y, m, d, msg)
   end
 
   def skip_windows!
-    skip_until 2020, 5, 1, "These have never passed" if windows?
+    skip_until 2020, 7, 1, "These have never passed" if windows?
```
Comment (james-stocks) [inline test/helper.rb]: This got merged in a separate PR so can be deleted from this PR

File: lib/inspec/input_registry.rb [hunk 7]
```diff
@@ -166,15 +166,44 @@ def bind_inputs_from_cli_args(profile_name, input_list)
...
+          value = YAML.load(value)
+        rescue Psych::SyntaxError => yaml_error
+          # It could be that we just tried to run JSON through the YAML parser.
+          require "json"
+          begin
+            value = JSON.parse(value)
+          rescue JSON::ParserError => json_error
+            msg = "ERROR: Unparseable value '#{value}' for --input #{input_name}.\n"
+            msg += "When treated as YAML, error: #{yaml_error.message}\n"
+            msg += "When treated as JSON, error: #{json_error.message}"
```
Comment (Schwad) [inline lib/inspec/input_registry.rb:202]: I like this approach/flow all the way down to showing possible errors for yaml/json.

Normally I'd be the first to ask conditionals to be named and refactored, along with nested begin/rescues, but in this case I actually think this design is the most straightforward/readable- nice!

File: test/helper.rb (hunk 6 above)
Comment (Schwad) [inline test/helper.rb]: I had to do this as well! But luckily with the merger I reverted and rebased master

File: lib/inspec/input_registry.rb (hunk 2 above)
Comment (clintoncwolfe) [reply lib/inspec/input_registry.rb:196]: After I wrote this, I realized that was probably true; do you think we should remove the second attempt for JSON?

**Comment by aaronlippold** [comment]
As a side note - another issue I will open in connection with this is to add pattern enforcement for inputs as well 

To start with - the easy cases 
Number [0-9]
String { /xxx.xxxx.xxxx.xxx/ }

The harder case would be JSON or more complex objects

I would assume there is a gem that does this well already

File: lib/inspec/input_registry.rb (hunk 2 above)
Comment (james-stocks) [reply lib/inspec/input_registry.rb:196]: I'm unsure, because the current code is expressive of what the method wants to do. Removing the JSON parsing attempt might cause confusion because someone will read this later and think "there's no JSON parsing code, how the heck did this ever work?!"
I'd suggest leaving it in. Possibly the comment on line 195 could be changed to reflect that `YAML.load` may or may not have been able to parse a JSON value and now we'll attempt a direct JSON parse

File: lib/inspec/input_registry.rb [hunk 8]
```diff
@@ -166,15 +166,47 @@ def bind_inputs_from_cli_args(profile_name, input_list)
...
+    def parse_cli_input_value(input_name, given_value)
+      value = given_value.chomp(",") # Trim trailing comma if any
+      case value
+      when /^true|false$/
+        value = value == "true"
+      when /^-?\d+$/
+        value = value.to_i
+      when /^-?\d+\.\d+$/
+        value = value.to_f
+      when /^(\[|\{).*(\]|\})$/
```
Comment (james-stocks) [inline lib/inspec/input_registry.rb:189]: Possibly this should match on at least one syntax-ish character e.g. `'` `"` `,` or `:`
That would allow something that's a simple string with brackets to continue as a literal string e.g. `[DEBUG]` or `{EXPECTED_CI_VARIABLE}`

Related to this, the `raise` on line 203 stops users from intending a string that matches whatever we use for the pattern here.
It should possibly continue as a literal string to avoid raising an error on something the user didn't intend to be parsed. Flip side to that would be users intending JSON/YAML but instead getting a string if (or _when_ for YAML :D ) they make a syntax error.

Comment (clintoncwolfe) [reply lib/inspec/input_registry.rb:189]: But `--input=[a,b,c]` is a valid YAML array; it's one of the test cases. It contains none of `",',:`.

I can change the raise to a warn, if that's blocking acceptance.

File: lib/inspec/input_registry.rb [hunk 9]
```diff
@@ -166,15 +166,47 @@ def bind_inputs_from_cli_args(profile_name, input_list)
...
         )
         find_or_register_input(input_name, profile_name, event: evt)
       end
     end
 
+    # Remove trailing commas, resolve type.
+    def parse_cli_input_value(input_name, given_value)
+      value = given_value.chomp(",") # Trim trailing comma if any
+      case value
+      when /^true|false$/
```
Comment (rmoles) [inline lib/inspec/input_registry.rb]: Would it be worth making this case insensitive to allow `True`,`TRUE` or `False`, `FALSE`?

File: lib/inspec/input_registry.rb (hunk 8 above)
Comment (james-stocks) [reply lib/inspec/input_registry.rb:189]: A warning feels better than an error, when the possibility is some edge case(s) with what the user's doing, thanks!

File: lib/inspec/input_registry.rb [hunk 10]
```diff
@@ -166,15 +166,47 @@ def bind_inputs_from_cli_args(profile_name, input_list)
...
+        require "yaml"
+        begin
+          value = YAML.load(value)
+        rescue Psych::SyntaxError => yaml_error
+          # It could be that we just tried to run JSON through the YAML parser.
+          require "json"
+          begin
+            value = JSON.parse(value)
+          rescue JSON::ParserError => json_error
+            msg = "ERROR: Unparseable value '#{value}' for --input #{input_name}.\n"
```
Comment (james-stocks) [inline lib/inspec/input_registry.rb]: The log method will already put the log level at the start of the string so `ERROR: ` can be omitted from the start of this string

**Review by james-stocks — APPROVED** [review]
This is great, thanks @clintoncwolfe 
I'll not merge immediately because there's an open question from Ross

**Comment by james-stocks** [comment]
Oops I forgot I left one more comment, so undo-ing my approval

File: lib/inspec/input_registry.rb (hunk 9 above)
Comment (clintoncwolfe) [reply lib/inspec/input_registry.rb]: OK

File: lib/inspec/input_registry.rb (hunk 10 above)
Comment (clintoncwolfe) [reply lib/inspec/input_registry.rb]: Done

**Review by james-stocks — APPROVED** [review]
This is awesome, thank you @clintoncwolfe :)

**Comment by KonradSchieban** [comment]
Thank you @clintoncwolfe !
